runtime/cudaq/builder/kernel_builder.cpp

```diff
@@ -619,6 +625,36 @@ QuakeValue qalloc(ImplicitLocOpBuilder &builder,
   return QuakeValue(builder, qubits);
 }
 
+QuakeValue qalloc(mlir::ImplicitLocOpBuilder &builder, cudaq::state *state,
+                  StateVectorStorage &stateVectorStorage) {
+  auto *context = builder.getContext();
+  auto index = stateVectorStorage.size();
+  stateVectorStorage.emplace_back(StateVectorVariant(state));
+  const char *getStatePtrCallback = "__nvqpp_getStatePtr";
+  {
+    auto parentModule =
+        builder.getBlock()->getParentOp()->getParentOfType<ModuleOp>();
+    IRBuilder irb(context);
+
+    if (failed(irb.loadIntrinsic(parentModule, getStatePtrCallback)))
+      throw std::runtime_error("loading callbacks should never fail");
+  }
+  auto statePtrTy = cudaq::cc::PointerType::get(cc::StateType::get(context));
+  std::intptr_t vecStor = reinterpret_cast<std::intptr_t>(&stateVectorStorage);
+  auto vecPtr = builder.create<arith::ConstantIntOp>(vecStor, 64);
+  auto idxOp = builder.create<arith::ConstantIntOp>(index, 64);
+  auto statePtr = builder
+                      .create<func::CallOp>(statePtrTy, getStatePtrCallback,
+                                            ValueRange{vecPtr, idxOp})
+                      .getResult(0);
```

Comment:
I think the newerish plan is to just pass to `State*` itself and not have a callback to unwrap the `State` and get a pointer to the vector data. Can you eliminate this call and pass `vecPtr` instead of this `statePtr` on line 654?

Reply:
Thanks for the pointer :) I've revised it to be much simpler in https://github.com/NVIDIA/cuda-quantum/pull/1722/commits/83c0c11a275750f69eaee0c4c4f8eab4d2abbe2d.